bionic/core/task_execution.py

```diff
@@ -147,65 +222,63 @@ def compute(self, task_key_logger):
             self.results_by_dnode = results_by_dnode
 
     def get_cached_results(self, task_key_logger):
-        "Returns the results of an already-completed entry."
+        "Returns the results of an already-computed entry."
 
         assert self.is_cached
 
         if self.results_by_dnode:
             return self.results_by_dnode
         return self.state.get_cached_results(task_key_logger)
 
+    def __str__(self):
+        return f"TaskRunnerEntry({', '.join(str(key) for key in self.state.task_keys)})"
+
 
 class EntryStage(Enum):
     """
-    Represents the stage of a `TaskRunnerEntry`.
+    Represents the stage of completion reached by a `TaskRunnerEntry`.
     """
 
     """
-    Entry was just created.
-    This is the always the first stage for an entry.
+    The entry is completed; we don't have any more work to do with it. All entries start
+    here (before any requests have been placed on them) and end here (once all the
+    requests have been satisfied).
+
     Valid next stages: [PENDING]
     """
-    NEW = auto()
+    COMPLETED = auto()
 
     """
-    Entry is waiting to be processed.
+    The entry is waiting to be processed. We know there's work to be done on it, but
```

Comment:
We create entries only when we need to process them. Since they start from `COMPLETED` state, we move them to `PENDING` to process them as soon as they are created.

How do you feel about making `PENDING` the starting stage? Or do you see a case in the future where we would create entries without knowing whether we need to process them? If not, I think it's better if we start from `PENDING` and make the change later on.

Reply:
> We create entries only when we need to process them.

I'm not sure this is true. We create an entry when we need to express some requirement on it, but that requirement might already be met. If the task state for that entry is already primed or cached, we don't need to actually put the entry into a `PENDING` state.

More abstractly: When we create a new entry, it has no requirements at all, so it's "completed" by definition, right?

Reply:
Duh, you are right. I was still thinking of it in terms of how we did it before. Also, you abstract explanation makes a lot of sense and convinced me easily!